avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java

```diff
@@ -362,28 +373,75 @@ Object read(Schema writerSchema) {
     }
 
     Object read(Schema writerSchema, Schema readerSchema) {
-      DatumReader<?> reader = getDatumReader(writerSchema, readerSchema);
-      int length = buffer.limit() - 1 - idSize;
-      if (writerSchema.getType().equals(Schema.Type.BYTES)) {
-        byte[] bytes = new byte[length];
-        buffer.get(bytes, 0, length);
-        return bytes;
-      } else {
-        int start = buffer.position() + buffer.arrayOffset();
-        try {
+      try {
+        DatumReader<?> reader = getDatumReader(writerSchema, readerSchema);
+        int length = buffer.limit() - 1 - idSize;
+        if (writerSchema.getType().equals(Schema.Type.BYTES)) {
+          byte[] bytes = new byte[length];
+          buffer.get(bytes, 0, length);
+          return bytes;
+        } else {
+          int start = buffer.position() + buffer.arrayOffset();
           Object result = reader.read(null, decoderFactory.binaryDecoder(buffer.array(),
               start, length, null));
           if (writerSchema.getType().equals(Schema.Type.STRING)) {
             return result.toString();
           } else {
             return result;
           }
-        } catch (IOException | RuntimeException e) {
-          // avro deserialization may throw AvroRuntimeException, NullPointerException, etc
-          throw new SerializationException("Error deserializing Avro message for id "
-              + schemaId, e);
         }
+      } catch (ExecutionException ex) {
+        throw new SerializationException("Error deserializing Avro message for id "
+            + schemaId, ex.getCause());
+      } catch (IOException | RuntimeException e) {
+        // avro deserialization may throw AvroRuntimeException, NullPointerException, etc
+        throw new SerializationException("Error deserializing Avro message for id "
+            + schemaId, e);
+      }
+    }
+  }
+
+  static class IdentityPair<K, V> {
+    private final K key;
+    private final V value;
+
+    public IdentityPair(K key, V value) {
+      this.key = key;
+      this.value = value;
+    }
+
+    public K getKey() {
+      return key;
+    }
+
+    public V getValue() {
+      return value;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
       }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      IdentityPair<?, ?> pair = (IdentityPair<?, ?>) o;
+      // Only perform identity check
+      return key == pair.key && value == pair.value;
+    }
+
+    @Override
+    public int hashCode() {
+      return System.identityHashCode(key) + System.identityHashCode(value);
```

Comment:
Not disagreeing, but curious why not the usual `Objects.hashCode(key, value)` ?

Reply:
@rayokota Did you consider a strong hash, like SHA256?

Reply:
This is an IdentityPair to only use identity (==) equality, therefore we use the identity hashCode.  The normal hashCode may be more expensive and the identity hashCode is sufficient.